tools/key_import/import.c

```diff
@@ -0,0 +1,767 @@
+        return 1;
+    }
+
+    return 0;
+}
+
+int
+c_get_info(void) {
+    CK_RV   rv;
+    CK_INFO info;
+
+    rv = C_GetInfo(&info);
+    if (rv != CKR_OK) {
+        PRINT_E("C_GetInfo has failed: %s", ckr_to_string(rv));
+        return 1;
+    }
+
+    return 0;
+}
+
+int
+c_get_slot(CK_SLOT_ID slot_id) {
+    CK_RV         rv;
+    CK_TOKEN_INFO info = { 0 };
+
+    rv = C_GetTokenInfo(slot_id, &info);
+    if (rv != CKR_OK) {
+        PRINT_E("C_GetTokenInfo has failed: %s", ckr_to_string(rv));
+        return 1;
+    }
+
+    if (info.flags & CKF_TOKEN_INITIALIZED) {
+        printf("Slot id: %ld\n", slot_id);
+
+        info.label[sizeof(info.label) - 1] = '\0';
+        printf("Token label: %s\n", info.label);
+
+        info.manufacturerID[sizeof(info.manufacturerID) - 1] = '\0';
+        printf("Token manufacturer id: %s\n", info.manufacturerID);
+
+        info.model[sizeof(info.model) - 1] = '\0';
+        printf("Token model: %s\n", info.model);
+
+        info.serialNumber[sizeof(info.serialNumber) - 1] = '\0';
+        printf("Token serial number: %s\n", info.serialNumber);
+    } else {
+        PRINT_E("Token is not initialized.");
+        return 1;
+    }
+
+    return 0;
+}
+
+int
+c_open_session(CK_SLOT_ID slot_id, CK_SESSION_HANDLE *session) {
+    CK_RV             rv;
+    CK_FLAGS          flags = CKF_SERIAL_SESSION | CKF_RW_SESSION;
+    CK_SESSION_HANDLE s = CK_INVALID_HANDLE;
+
+    rv = C_OpenSession(slot_id, flags, NULL, NULL, &s);
+    if (rv != CKR_OK) {
+        PRINT_E("C_OpenSession has failed: %s", ckr_to_string(rv));
+        return 1;
+    }
+
+    *session = s;
+
+    return 0;
+}
+
+int
+c_close_session(CK_SESSION_HANDLE session) {
+    CK_RV rv;
+
+    if (session == CK_INVALID_HANDLE) {
+        return 0;
+    }
+
+    rv = C_CloseSession(session);
+    if (rv != CKR_OK) {
+        PRINT_E("C_CloseSession has failed: %s", ckr_to_string(rv));
+        return 1;
+    }
+
+    return 0;
+}
+
+int
+c_login(CK_SESSION_HANDLE session, CK_UTF8CHAR_PTR user_pin) {
+    CK_RV        rv;
+    CK_USER_TYPE user_type = CKU_USER;
+    CK_ULONG     user_pin_len = 0;
+
+    if (session == CK_INVALID_HANDLE) {
+        return 0;
+    }
+
+    if (user_pin) {
+        user_pin_len = strlen((char *)user_pin);
+    }
+
+    rv = C_Login(session, user_type, user_pin, user_pin_len);
+    if (rv != CKR_OK) {
+        PRINT_E("C_Login has failed: %s", ckr_to_string(rv));
+        return 1;
+    }
+
+    return 0;
+}
+
+int
+c_generate_keypair(CK_SESSION_HANDLE session,
+                   CK_MECHANISM     *mech,
+                   CK_ATTRIBUTE     *pub,
+                   CK_ULONG          pub_len,
+                   CK_ATTRIBUTE     *priv,
+                   CK_ULONG          priv_len) {
+    CK_RV            rv;
+    CK_OBJECT_HANDLE pubkey;
+    CK_OBJECT_HANDLE privkey;
+
+    rv = C_GenerateKeyPair(session, mech, pub, pub_len, priv, priv_len, &pubkey, &privkey);
```

Comment:
I do believe this should have been C_CreateObject called twice, once for the pub key and once for the private key as this doesn't generate a key pair.